File: packages/agentenkit/test/run-follow.test.ts
import { describe, expect, it } from 'bun:test';
import { simulateReadableStream } from 'ai';
import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import { MockLanguageModelV1 } from 'ai/test';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryRunStreams, MemoryStorage } from '../src/adapters/memory.js';
import { followFrame, formatCursor, parseCursor, type FollowFrame } from '../src/core/follow.js';
import { openSegment } from '../src/core/segment.js';
import { resolveConfig } from '../src/core/types.js';

// Workstream S6: a follow reads the thread record and its run streams as one
// sequence. The Go package runs the same cases under the same names
// (run_follow_test.go).

function textModel(...replies: string[]) {
  let n = 0;
  return new MockLanguageModelV1({
    provider: 'mock',
    modelId: 'mock',
    doStream: async () => {
      const chunks: LanguageModelV1StreamPart[] = [
        { type: 'text-delta', textDelta: replies[Math.min(n++, replies.length - 1)]! },
        { type: 'finish', finishReason: 'stop', usage: { promptTokens: 10, completionTokens: 5 } },
      ];
      return { stream: simulateReadableStream({ chunks }), rawCall: { rawPrompt: null, rawSettings: {} } };
    },
  });
}

async function makeRuntime(model: any) {
  const queue = new MemoryQueue();
  const bus = new MemoryBus();
  const streams = new MemoryRunStreams();
  const runtime = await setupAgentCore({
    storage: new MemoryStorage(), admin: new MemoryAdminStore(), bus, queue, kv: new MemoryKv(), streams,
    resolveModel: () => ({ instance: () => model, contextWindow: 128_000 }),
    config: resolveConfig({ streamFlushMs: 0 }),
  });
  const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
  return { runtime, queue, bus, streams, chat };
}

/** Read frames until `done` holds, or fail after a second. */
async function readUntil(gen: AsyncGenerator<FollowFrame>, done: (frames: FollowFrame[]) => boolean) {
  const frames: FollowFrame[] = [];
  const timeout = setTimeout(() => void gen.return(undefined as never), 1000);
  for await (const frame of gen) {
    frames.push(frame);
    if (done(frames)) break;
  }
  clearTimeout(timeout);
  await gen.return(undefined as never);
  return frames;
}

const texts = (frames: FollowFrame[]) =>
  frames.flatMap((f) => (f.kind === 'stream' && f.item.type === 'TEXT_MESSAGE_CONTENT' ? [f.item.delta] : []));
const ends = (frames: FollowFrame[]) =>
  frames.filter((f) => f.kind === 'stream' && f.item.type === 'RUN_FINISHED').length;

describe('run follow', () => {
  it("follow moves to the next run's stream on its start notice", async () => {
    const r = await makeRuntime(textModel('first', 'second'));
    const ran = await r.chat.run({ prompt: 'a' });
    const gen = r.runtime.events.follow(ran.threadId);
    const reading = readUntil(gen, (f) => ends(f) === 2);
    await new Promise((resolve) => setTimeout(resolve, 10));
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    await r.chat.run({ threadId: ran.threadId, prompt: 'b' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const frames = await reading;
    expect(texts(frames)).toEqual(['first', 'second']);
    const streams = [...new Set(frames.flatMap((f) => (f.kind === 'stream' ? [f.streamId] : [])))];
    expect(streams).toHaveLength(2); // one per run
  });

  it('a reconnect with Last-Event-ID resumes inside the stream', async () => {
    const r = await makeRuntime(textModel('x'));
    const seg = (await openSegment(r.runtime.ports(), 't1', 'run1'))!;
    await seg.push([
      { type: 'TEXT_MESSAGE_START', messageId: 'm1', role: 'assistant' },
      { type: 'TEXT_MESSAGE_CONTENT', messageId: 'm1', delta: 'seen ' },
    ]);
    const [, , seen] = (await r.streams.snapshot('run1:1'))!.items;
    await seg.push([{ type: 'TEXT_MESSAGE_CONTENT', messageId: 'm1', delta: 'new' }]);
    await seg.close({ type: 'RUN_FINISHED', status: 'finished' });

    // The browser sends back the id of the last frame it got.
    const gen = r.runtime.events.follow('t1', { cursor: `-1 run1:1 ${seen!.offset}` });
    const frames = await readUntil(gen, (f) => ends(f) === 1);
    expect(texts(frames)).toEqual(['new']);
  });

  it('a reconnect to a gone stream gets one SNAPSHOT with only the newer messages, on the same connection', async () => {
    const r = await makeRuntime(textModel('one', 'two'));
    const ran = await r.chat.run({ prompt: 'a' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const had = (await r.runtime.getThreadSnapshot(ran.threadId))!.messages.at(-1)!.id;
    const second = await r.chat.run({ threadId: ran.threadId, prompt: 'b' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    await r.streams.delete(`${ran.runId}:1`); // past its grace window

    const gen = r.runtime.events.follow(ran.threadId, {
      cursor: { seq: 0, streamId: `${ran.runId}:1`, offset: '1' },
      lastMessageId: had,
    });
    const frames = await readUntil(gen, (f) => f.some((x) => x.kind === 'snapshot'));
    const snap = frames.find((f) => f.kind === 'snapshot');
    expect(snap).toBeDefined();
    if (snap?.kind !== 'snapshot') throw new Error('no snapshot');
    expect(snap.snapshot.messages.map((m) => m.role)).toEqual(['user', 'assistant']); // run b's only
    expect(snap.snapshot.stream?.streamId).toBe(`${second.runId}:1`);
  });

  it("a cursor naming another thread's stream is never read", async () => {
    const r = await makeRuntime(textModel('secret', 'mine'));
    const other = await r.chat.run({ prompt: 'a' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const own = await r.chat.run({ prompt: 'b' }); // a thread of its own
    await r.runtime.worker.handleJob(r.queue.items.shift()!);

    const gen = r.runtime.events.follow(own.threadId, { cursor: `0 ${other.runId}:1 1` });
    const frames = await readUntil(gen, (f) => f.some((x) => x.kind === 'snapshot'));
    expect(texts(frames)).not.toContain('secret');
    const snap = frames.find((f) => f.kind === 'snapshot');
    if (snap?.kind !== 'snapshot') throw new Error('no snapshot');
    expect(snap.snapshot.thread.id).toBe(own.threadId); // its own thread instead
  });

  it('the sse id carries the record seq and the stream position', () => {
    const at = { seq: 4 };
    const thread = followFrame(
      { kind: 'thread', event: { threadId: 't', seq: 5, type: 'X', payload: null, createdAt: new Date(0) } }, at,
    );
    expect(thread.startsWith('id: 5 - -\n')).toBe(true);
    const item = followFrame(
      { kind: 'stream', streamId: 'r1:2', item: { type: 'TEXT_MESSAGE_CONTENT', messageId: 'm', delta: 'd', offset: '9' } }, at,
    );
    expect(item.startsWith('id: 5 r1:2 9\n')).toBe(true);
    const notice = followFrame(
      { kind: 'thread', event: { threadId: 't', seq: 0, type: 'HEARTBEAT', payload: null, createdAt: new Date(0) } }, at,
    );
    expect(notice).not.toContain('id:'); // leaves the browser's cursor as it was
  });

  it('a cursor reads its wire form and a bare seq', () => {
    expect(parseCursor('5 r1:2 9')).toEqual({ seq: 5, streamId: 'r1:2', offset: '9' });
    expect(parseCursor('5 - -')).toEqual({ seq: 5 });
    expect(parseCursor('12')).toEqual({ seq: 12 });
    expect(parseCursor('junk')).toBeNull();
    expect(formatCursor({ seq: 3, streamId: 'r:1', offset: '1-0' })).toBe('3 r:1 1-0');
  });

  it('stream content does not go on the bus while a segment is open', async () => {
    const r = await makeRuntime(textModel('hello'));
    const ran = await r.chat.run({ prompt: 'a' });
    await r.runtime.worker.handleJob(r.queue.items.shift()!);
    const onBus = r.bus.published.filter((e) => e.threadId === ran.threadId).map((e) => e.type);
    expect(onBus).not.toContain('CHUNK');
    expect(onBus).not.toContain('STEP_COMMITTED');
    expect(onBus).toContain('STATE_CHANGE'); // thread notices still go out
    expect(onBus).toContain('RUN_STARTED');
  });
});

File: packages/go-agentenkit/run_follow_test.go
package agentenkit_test

import (
	"context"
	"strings"
	"testing"
	"time"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/core"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
)

// Workstream S6: a follow reads the thread record and its run streams as
// one sequence. The TS package runs the same cases under the same names
// (test/run-follow.test.ts).

// readUntil reads frames until done holds, or fails after a second.
func readUntil(t *testing.T, stream *core.FrameStream, done func([]core.FollowFrame) bool) []core.FollowFrame {
	t.Helper()
	var frames []core.FollowFrame
	timeout := time.After(time.Second)
	for !done(frames) {
		select {
		case f, ok := <-stream.Frames():
			if !ok {
				t.Fatal("follow closed")
			}
			frames = append(frames, f)
		case <-timeout:
			t.Fatalf("timed out; got %d frames", len(frames))
		}
	}
	return frames
}

func frameTexts(frames []core.FollowFrame) []string {
	var out []string
	for _, f := range frames {
		if f.Kind == core.FrameKindStream {
			if c, ok := f.Item.Event.(*ports.TextMessageContentEvent); ok {
				out = append(out, c.Delta)
			}
		}
	}
	return out
}

func frameEnds(frames []core.FollowFrame) int {
	n := 0
	for _, f := range frames {
		if f.Kind == core.FrameKindStream {
			if _, ok := f.Item.Event.(*ports.RunFinishedEvent); ok {
				n++
			}
		}
	}
	return n
}

func TestRunFollow(t *testing.T) {
	t.Run("follow moves to the next run's stream on its start notice", func(t *testing.T) {
		h, _ := streamRuntime(t, scripted(step{text: "first"}, step{text: "second"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
		ran := h.run(t, chat, agentenkit.RunInput{Prompt: "a"})
		ctx, cancel := context.WithCancel(h.ctx)
		defer cancel()
		stream, err := h.rt.Events.Follow(ctx, ran.ThreadID, agentenkit.FollowStateOptions{})
		must(t, err)
		time.Sleep(10 * time.Millisecond)
		h.handleNext(t)
		h.run(t, chat, agentenkit.RunInput{ThreadID: ran.ThreadID, Prompt: "b"})
		h.handleNext(t)
		frames := readUntil(t, stream, func(f []core.FollowFrame) bool { return frameEnds(f) == 2 })
		mustStrings(t, frameTexts(frames), []string{"first", "second"}, "both runs' text")
		streams := map[string]bool{}
		for _, f := range frames {
			if f.Kind == core.FrameKindStream {
				streams[f.StreamID] = true
			}
		}
		mustEqual(t, len(streams), 2, "one stream per run")
	})

	t.Run("a reconnect with Last-Event-ID resumes inside the stream", func(t *testing.T) {
		h, streams := streamRuntime(t, scripted(step{text: "x"}))
		seg := core.OpenSegment(h.ctx, h.rt.Ports(nil), "t1", "run1")
		seg.Push(h.ctx,
			&ports.TextMessageStartEvent{MessageID: "m1", Role: "assistant"},
			&ports.TextMessageContentEvent{MessageID: "m1", Delta: "seen "})
		seen := streamItems(t, streams, "run1:1")[2]
		seg.Push(h.ctx, &ports.TextMessageContentEvent{MessageID: "m1", Delta: "new"})
		seg.Close(h.ctx, &ports.RunFinishedEvent{Status: "finished"})

		// The browser sends back the id of the last frame it got.
		ctx, cancel := context.WithCancel(h.ctx)
		defer cancel()
		stream, err := h.rt.Events.Follow(ctx, "t1", agentenkit.FollowStateOptions{Cursor: "-1 run1:1 " + seen.Offset})
		must(t, err)
		frames := readUntil(t, stream, func(f []core.FollowFrame) bool { return frameEnds(f) == 1 })
		mustStrings(t, frameTexts(frames), []string{"new"}, "only what came after")
	})

	t.Run("a reconnect to a gone stream gets one SNAPSHOT with only the newer messages, on the same connection", func(t *testing.T) {
		h, streams := streamRuntime(t, scripted(step{text: "one"}, step{text: "two"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
		ran := h.run(t, chat, agentenkit.RunInput{Prompt: "a"})
		h.handleNext(t)
		snap, err := h.rt.GetThreadSnapshot(h.ctx, ran.ThreadID, nil)
		must(t, err)
		had := snap.Messages[len(snap.Messages)-1].ID
		second := h.run(t, chat, agentenkit.RunInput{ThreadID: ran.ThreadID, Prompt: "b"})
		h.handleNext(t)
		must(t, streams.Delete(h.ctx, ran.RunID+":1")) // past its grace window

		ctx, cancel := context.WithCancel(h.ctx)
		defer cancel()
		stream, err := h.rt.Events.Follow(ctx, ran.ThreadID, agentenkit.FollowStateOptions{
			Cursor: "0 " + ran.RunID + ":1 1", LastMessageID: had,
		})
		must(t, err)
		frames := readUntil(t, stream, func(f []core.FollowFrame) bool {
			for _, x := range f {
				if x.Kind == core.FrameKindSnapshot {
					return true
				}
			}
			return false
		})
		var got *ports.ThreadSnapshot
		for _, f := range frames {
			if f.Kind == core.FrameKindSnapshot {
				got = f.Snapshot
			}
		}
		var roles []string
		for _, m := range got.Messages {
			roles = append(roles, string(m.Role))
		}
		mustStrings(t, roles, []string{"user", "assistant"}, "run b's messages only")
		mustEqual(t, got.Stream.StreamID, second.RunID+":1", "and run b's stream")
	})

	t.Run("the sse id carries the record seq and the stream position", func(t *testing.T) {
		at := core.ThreadCursor{Seq: 4}
		thread := core.FollowFrameSSE(core.FollowFrame{Kind: core.FrameKindThread, Event: &ports.AgentEvent{ThreadID: "t", Seq: 5, Type: "X"}}, &at)
		if !strings.HasPrefix(thread, "id: 5 - -\n") {
			t.Fatalf("thread frame: %q", thread)
		}
		item := core.FollowFrameSSE(core.FollowFrame{Kind: core.FrameKindStream, StreamID: "r1:2", Item: &ports.StreamItem{
			Offset: "9", Event: &ports.TextMessageContentEvent{MessageID: "m", Delta: "d"},
		}}, &at)
		if !strings.HasPrefix(item, "id: 5 r1:2 9\n") {
			t.Fatalf("stream frame: %q", item)
		}
		notice := core.FollowFrameSSE(core.FollowFrame{Kind: core.FrameKindThread, Event: &ports.AgentEvent{ThreadID: "t", Type: "HEARTBEAT"}}, &at)
		if strings.Contains(notice, "id:") {
			t.Fatalf("a notice leaves the browser's cursor as it was: %q", notice)
		}
	})

	t.Run("a cursor reads its wire form and a bare seq", func(t *testing.T) {
		c, ok := core.ParseCursor("5 r1:2 9")
		mustEqual(t, ok && c == core.ThreadCursor{Seq: 5, StreamID: "r1:2", Offset: "9"}, true, "full")
		c, ok = core.ParseCursor("5 - -")
		mustEqual(t, ok && c == core.ThreadCursor{Seq: 5}, true, "record only")
		c, ok = core.ParseCursor("12")
		mustEqual(t, ok && c == core.ThreadCursor{Seq: 12}, true, "bare seq")
		_, ok = core.ParseCursor("junk")
		mustEqual(t, ok, false, "junk")
		mustEqual(t, core.FormatCursor(core.ThreadCursor{Seq: 3, StreamID: "r:1", Offset: "1-0"}), "3 r:1 1-0", "format")
	})

	t.Run("stream content does not go on the bus while a segment is open", func(t *testing.T) {
		h, _ := streamRuntime(t, scripted(step{text: "hello"}))
		chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
		ran := h.run(t, chat, agentenkit.RunInput{Prompt: "a"})
		h.handleNext(t)
		var onBus []string
		for _, e := range h.events(ran.ThreadID, "") {
			onBus = append(onBus, e.Type)
		}
		joined := strings.Join(onBus, " ")
		if strings.Contains(joined, "CHUNK") || strings.Contains(joined, "STEP_COMMITTED") {
			t.Fatalf("stream content on the bus: %v", onBus)
		}
		if !strings.Contains(joined, "STATE_CHANGE") || !strings.Contains(joined, "RUN_STARTED") {
			t.Fatalf("thread notices missing: %v", onBus)
		}
	})
}
